subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -0,0 +1,99 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.recomp;
+
+import com.google.common.collect.Sets;
+import org.gradle.internal.change.DefaultFileChange;
+import org.gradle.internal.file.FileType;
+import org.gradle.internal.fingerprint.impl.IgnoredPathFingerprintingStrategy;
+import org.gradle.work.FileChange;
+import org.gradle.work.InputChanges;
+
+import java.io.File;
+import java.util.Set;
+
+public class GroovyRecompilationSpecProvider implements RecompilationSpecProvider {
+    private final InputChanges inputChanges;
+    private final Iterable<FileChange> classpathChanges;
+    private final Iterable<FileChange> sourceChanges;
+    private final SourceFileClassNameConverter sourceFileClassNameConverter;
+
+    public GroovyRecompilationSpecProvider(InputChanges inputChanges,
+                                           Iterable<FileChange> sourceChanges,
+                                           Iterable<FileChange> classpathChanges,
+                                           SourceFileClassNameConverter sourceToNameConverter) {
+        this.inputChanges = inputChanges;
+        this.classpathChanges = classpathChanges;
+        this.sourceChanges = sourceChanges;
+        this.sourceFileClassNameConverter = sourceToNameConverter;
+    }
+
+    @Override
+    public boolean isIncremental() {
+        return inputChanges.isIncremental();
+    }
+
+    @Override
+    public RecompilationSpec provideRecompilationSpec(CurrentCompilation current, PreviousCompilation previous) {
+        RecompilationSpec spec = new RecompilationSpec(sourceFileClassNameConverter);
+
+        processClasspathChanges(current, previous, spec);
+        processOtherChanges(current, previous, spec);
+
+        spec.getClassesToProcess().addAll(previous.getTypesToReprocess());
+        return spec;
+    }
+
+    private void processClasspathChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec) {
+        ClasspathEntryChangeProcessor classpathEntryChangeProcessor = new ClasspathEntryChangeProcessor(current.getClasspathSnapshot(), previous);
+        Set<File> transformedClasspathEntryDetectionSet = Sets.newHashSet();
+        for (FileChange fileChange : classpathChanges) {
+            switch (fileChange.getChangeType()) {
+                case ADDED:
+                    if (!transformedClasspathEntryDetectionSet.add(fileChange.getFile())) {
+                        spec.setFullRebuildCause("Classpath has been changed", null);
+                        return;
+                    }
+                    DefaultFileChange added = DefaultFileChange.added(fileChange.getFile().getAbsolutePath(), "classpathEntry", FileType.RegularFile, IgnoredPathFingerprintingStrategy.IGNORED_PATH);
```

Comment:
Why do you need to create a new `FileChange` here? Isn't `fileChange` the thing you should process?

Note that the file change may be a class file in a directory - then the `fileChange.normalizedPath != ""`. I don't know exactly if `classpathEntryChangeProcessor` can deal with this.

I looked a little bit into it and it seems like the `ClasspathChangeDependentsFinder` will need to be adjusted or we need to gather all the changes for one root until the next root shows up and pass it gathered to the `classpathEntryChangeProcessor`.

It would be great if we wouldn't need to keep track of classes in directories at all and use the `InputChanges` directly - I think this is possible by handling non-root files (i.e. `normalizedPath != ""`) differently.